special-episode-1/step9/app/js/lib/app.js

```diff
@@ -0,0 +1,82 @@
+import $ from 'jquery';
+import Web3 from 'web3';
+import TruffleContract from 'truffle-contract';
+import artifact from '../../../contracts/ToDo.sol';
+import { renderTasks } from './render';
+import { getAccount, getTasks } from './actions';
+
+class App {
+  constructor(config) {
+    this.config = config;
+  }
+
+  setup() {
+    const { ethereumUrl } = this.config;
+    const web3 = new Web3(new Web3.providers.HttpProvider(ethereumUrl));
+
+    const Todo = new TruffleContract(artifact);
+    Todo.setProvider(web3.currentProvider);
+
+    const networks = Object.keys(artifact.networks);
+    const network = networks[networks.length - 1];
+    const address = artifact.networks[network].address;
+
+    this.web3 = web3;
+    this.address = address;
+    this.Todo = Todo;
+    this.$tasks = $('#tasks');
+    this.$newTask = $('#new-task');
+    this.$taskContent = $('#task-content');
+    this.$taskAuthor = $('#task-author');
+
+    return new Promise((resolve, reject) => {
+      getAccount(this.web3)
+        .then((account) => {
+          this.account = account;
+          return Todo.at(address);
+        })
+        .then((todo) => {
+          this.todo = todo;
+          resolve(todo);
+        })
+        .catch((error) => {
+          reject(error);
+        });
+    });
+  }
+
+  getAndRenderTasks() {
+    getTasks(this.todo)
+      .then((tasks) => {
+        renderTasks(this.$tasks, tasks);
+        tasks.forEach((task, i) => {
+          i += 1
+          $(`#checkbox-${i}`).on('change', (event) => {
```

Comment:
I think this could be done only once in the `init()` function by listening to click events on a top DOM node(event bubbling). See this:

https://github.com/jklepatch/eattheblocks/blob/master/special-episode-1/finished/app/js/lib/events.js

Reply:
done! i forgot about event delegation,,,, been a while since I've used jquery :]